design/multi-homed-cluster.md

```diff
@@ -0,0 +1,149 @@
+network, cluster CRD needs to tell the network provider the appropiate
+`NetworkInterfaceSelector`. `NetworkInterfaceSelector` will be provided as list
+of `interfaces` key-value.
+
+```yaml
+network:
+  provider: <network-provider>
+  interfaces:
+    <key>: <network-interface-selector>
+    <key>: <network-interface-selector>
+```
+
+### Network Provider
+
+Network provider determines multi-homing method and the network interface
+selector. Using host network provider, pod can use host's network namespace
+and its interfaces to use as cluster network. Using popular multi-plugin CNI
+plugin such as [multus][multus-cni], Rook operators can attach multiple network
+interfaces to pods.
+
+In the future, works can be added to support other multi network CNI plugins
+such as [knitter][knitter-cni] or [genie][genie-cni].
+
+One thing to remember, leaving the network configuration empty will default to
+kubernetes cluster default networking.
+
+### Keys
+
+The key for each interface key-value pairs are left to each storage providers to
+decide. Having network interface selector with key value `server` and `broker`
+makes more sense for EdgeFS storage provider while Ceph will have `public` and
+`cluster`.
+
+### Network Interface Selector
+
+Network interface selector is fed to network provider to connect pod to cluster
+network. This selector may vary from one network provider to another. For
+example, host network provider only needs to know the interface name.
+
+On the other hand, multi-plugin CNI plugin needs to know the network attachment
+definition's name and vice versa. Multi-plugin such as multus may seem to follow
+_Network Attachment Selection Annotation_ documented at [Kubernetes Network
+Custom Resource Definition De-facto Standard][network-crd]. However, they only
+implement a part of the standard and have extra features not covered by the
+standard such as `<namespace>/` notation, `@<interface-name>` notation or
+`interfaceRequest` property documented [here][multus-annotation].
+
+## Example Configurations
+
+### EdgeFS Rook Cluster Network definition example
+
+1. Default networking
+
+```yaml
+# network:
+#   provider: <network-provider>
+#   interfaces:
+#     <key>: <network-interface-selector>
+#     <key>: <network-interface-selector>
+```
+
+2. Host network provider
+
+```yaml
+network:
+  provider: host
+  interfaces:
+    server: eth0
+    broker: eth0
+```
+
+3. Multus network provider
+
+```yaml
+network:
+  provider: multus
+  interfaces:
+    server: flannel # defaults to net1 interface
+    broker: flannel # defaults to net2 interface
+```
+
+The issue with this configuration is that `netx` interface name is given in the
+order they are applied by multus. No sure way to know if it means both keys to
+use the same network and interface, or in different interfaces just by the
+configuration alone.
+
+4. Multus network provider with JSON selector
+
+```yaml
+network:
+  provider: multus
+  interfaces:
+    server: '{
```

Comment:
What is this JSON based on? Why not create a spec for this?

Reply:
The serialization will typically work for both yaml and json. For consistency it would be preferred to use yaml in the examples. Will this not be translated into the [types.go](https://github.com/rook/rook/blob/master/pkg/apis/edgefs.rook.io/v1beta1/types.go#L55) structs? Or does it need to be freeform to pass through to the network provider (multus in this case)?

Reply:
I am myself more inclined to the freeform approach. CNI configuration and the original annotations standard are already in json.

This issue is even quite messy there https://github.com/intel/multus-cni/issues/294.